File: packages/api/src/models/BookClub.ts
/**
 * BookClub Model — Reading groups / book clubs
 */

import { DataTypes, Model, Optional } from 'sequelize';
import { sequelize } from '../db';

// ---------------------------------------------------------------------------
// BookClub
// ---------------------------------------------------------------------------

interface BookClubAttributes {
  id: string;
  name: string;
  description?: string;
  coverImage?: string;
  createdBy: string;
  currentBookId?: string;
  isPrivate: boolean;
  inviteCode: string;
  maxMembers: number;
  createdAt: Date;
  updatedAt: Date;
}

interface BookClubCreationAttributes extends Optional<BookClubAttributes, 'id' | 'description' | 'coverImage' | 'currentBookId' | 'isPrivate' | 'inviteCode' | 'maxMembers' | 'createdAt' | 'updatedAt'> {}

export class BookClub extends Model<BookClubAttributes, BookClubCreationAttributes> implements BookClubAttributes {
  public id!: string;
  public name!: string;
  public description?: string;
  public coverImage?: string;
  public createdBy!: string;
  public currentBookId?: string;
  public isPrivate!: boolean;
  public inviteCode!: string;
  public maxMembers!: number;
  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

function generateInviteCode(): string {
  const chars = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';
  let code = '';
  for (let i = 0; i < 6; i++) {
    code += chars[Math.floor(Math.random() * chars.length)];
  }
  return code;
}

BookClub.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    name: {
      type: DataTypes.STRING,
      allowNull: false,
      validate: { len: [1, 100] },
    },
    description: {
      type: DataTypes.STRING,
      allowNull: true,
      validate: { len: [0, 500] },
    },
    coverImage: {
      type: DataTypes.STRING,
      allowNull: true,
    },
    createdBy: {
      type: DataTypes.UUID,
      allowNull: false,
      references: { model: 'users', key: 'id' },
      onDelete: 'CASCADE',
    },
    currentBookId: {
      type: DataTypes.UUID,
      allowNull: true,
      references: { model: 'books', key: 'id' },
      onDelete: 'SET NULL',
    },
    isPrivate: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false,
    },
    inviteCode: {
      type: DataTypes.STRING(6),
      allowNull: false,
      defaultValue: () => generateInviteCode(),
      unique: true,
    },
    maxMembers: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 20,
      validate: { min: 2, max: 100 },
    },
    createdAt: { type: DataTypes.DATE, allowNull: false },
    updatedAt: { type: DataTypes.DATE, allowNull: false },
  },
  {
    sequelize,
    tableName: 'book_clubs',
    timestamps: true,
    indexes: [
      { fields: ['created_by'] },
      { fields: ['invite_code'], unique: true },
      { fields: ['current_book_id'] },
    ],
  }
);

// ---------------------------------------------------------------------------
// BookClubMember
// ---------------------------------------------------------------------------

export type ClubRole = 'admin' | 'moderator' | 'member';

interface BookClubMemberAttributes {
  id: string;
  clubId: string;
  userId: string;
  role: ClubRole;
  joinedAt: Date;
}

interface BookClubMemberCreationAttributes extends Optional<BookClubMemberAttributes, 'id' | 'role' | 'joinedAt'> {}

export class BookClubMember extends Model<BookClubMemberAttributes, BookClubMemberCreationAttributes> implements BookClubMemberAttributes {
  public id!: string;
  public clubId!: string;
  public userId!: string;
  public role!: ClubRole;
  public joinedAt!: Date;
}

BookClubMember.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    clubId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: { model: 'book_clubs', key: 'id' },
      onDelete: 'CASCADE',
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: { model: 'users', key: 'id' },
      onDelete: 'CASCADE',
    },
    role: {
      type: DataTypes.STRING(20),
      allowNull: false,
      defaultValue: 'member',
      validate: { isIn: [['admin', 'moderator', 'member']] },
    },
    joinedAt: {
      type: DataTypes.DATE,
      allowNull: false,
      defaultValue: DataTypes.NOW,
    },
  },
  {
    sequelize,
    tableName: 'book_club_members',
    timestamps: false,
    indexes: [
      { fields: ['club_id'] },
      { fields: ['user_id'] },
      { fields: ['club_id', 'user_id'], unique: true },
    ],
  }
);

// ---------------------------------------------------------------------------
// Associations
// ---------------------------------------------------------------------------

BookClub.hasMany(BookClubMember, { foreignKey: 'clubId', as: 'members' });
BookClubMember.belongsTo(BookClub, { foreignKey: 'clubId', as: 'club' });

BookClub.belongsTo(BookClubMember, { foreignKey: 'createdBy', as: 'creator', targetKey: 'userId', constraints: false });

File: packages/server/app/models/book_club.py
"""Book club models."""

import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Optional
from uuid import UUID

from sqlalchemy import Boolean, ForeignKey, Index, Integer, String, Text, UniqueConstraint, text
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base

if TYPE_CHECKING:
    from app.models.book import Book
    from app.models.user import User


class BookClub(Base):
    __tablename__ = 'book_clubs'
    __table_args__ = (
        Index('ix_book_clubs_invite_code', 'invite_code', unique=True),
    )

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        server_default=text('gen_random_uuid()'),
    )
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    cover_image: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    created_by: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    current_book_id: Mapped[Optional[UUID]] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('books.id', ondelete='SET NULL'),
        nullable=True,
        index=True,
    )
    is_private: Mapped[bool] = mapped_column(Boolean, default=False)
    invite_code: Mapped[str] = mapped_column(
        String(6),
        unique=True,
        nullable=False,
        server_default=text("upper(substring(md5(random()::text), 1, 6))"),
    )
    max_members: Mapped[int] = mapped_column(Integer, default=20)
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    creator: Mapped['User'] = relationship(
        'User',
        back_populates='created_clubs',
        foreign_keys=[created_by],
    )
    current_book: Mapped[Optional['Book']] = relationship('Book')
    members: Mapped[list['BookClubMember']] = relationship(
        'BookClubMember',
        back_populates='club',
        cascade='all, delete-orphan',
    )
    discussions: Mapped[list['ClubDiscussion']] = relationship(
        'ClubDiscussion',
        back_populates='club',
        cascade='all, delete-orphan',
    )


class BookClubMember(Base):
    __tablename__ = 'book_club_members'
    __table_args__ = (
        UniqueConstraint('club_id', 'user_id', name='uq_club_member'),
    )

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        server_default=text('gen_random_uuid()'),
    )
    club_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('book_clubs.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False,
        index=True,
    )
    role: Mapped[str] = mapped_column(String(20), default='member')
    joined_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)

    club: Mapped['BookClub'] = relationship('BookClub', back_populates='members')
    user: Mapped['User'] = relationship('User', back_populates='club_memberships')


class ClubDiscussion(Base):
    __tablename__ = 'club_discussions'
    __table_args__ = (
        Index('ix_club_discussions_club_created', 'club_id', 'created_at'),
    )

    id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        server_default=text('gen_random_uuid()'),
    )
    club_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('book_clubs.id', ondelete='CASCADE'),
        nullable=False,
    )
    user_id: Mapped[UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey('users.id', ondelete='CASCADE'),
        nullable=False,
    )
    content: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)

    club: Mapped['BookClub'] = relationship('BookClub', back_populates='discussions')
    user: Mapped['User'] = relationship('User')
